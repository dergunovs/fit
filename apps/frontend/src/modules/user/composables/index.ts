import { computed, Ref } from 'vue';

import { useTI18n } from '@/common/composables';

export function useUserFormTabs(isEdit: boolean, isAdmin: Ref<boolean>) {
  const { t } = useTI18n();

  const generalTab = computed(() => {
    return { value: 'general', title: t('user.general') };
  });

  const exercisesTab = computed(() => {
    return { value: 'exercises', title: t('exercise.many') };
  });

  const otherTabs = computed(() => [
    { value: 'goals', title: t('goals') },
    { value: 'weight', title: t('weight') },
    { value: 'templates', title: t('template.many') },
  ]);

  const TABS = computed(() => {
    if (!isEdit) return [generalTab.value];
    if (isAdmin.value) return [generalTab.value, ...otherTabs.value];

    return [generalTab.value, ...otherTabs.value, exercisesTab.value];
  });

  return { TABS };
}
