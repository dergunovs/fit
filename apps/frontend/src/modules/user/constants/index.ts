export const URL_USER = '/users';
export const URL_USER_CREATE = `${URL_USER}/create`;
export const URL_USER_EDIT = `${URL_USER}/edit`;
export const URL_USER_PROFILE = `${URL_USER}/profile`;

export const USER_GOALS_OPTIONS = {
  activities: [1, 2, 3, 4, 5],
  sets: [12, 16, 20, 24, 30],
  repeats: [8, 10, 12, 15, 20],
  duration: [20, 30, 40, 50, 60],
};
