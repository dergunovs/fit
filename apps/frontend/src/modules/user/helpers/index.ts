import { IEquipment, IUserEquipment } from 'fitness-tracker-contracts';

export function excludeChoosenUserEquipment(equipments: IEquipment[], userEquipments?: IUserEquipment[]) {
  return equipments.filter(
    (equipment) => !userEquipments?.some((equipmentToFilter) => equipment._id === equipmentToFilter.equipment?._id)
  );
}
