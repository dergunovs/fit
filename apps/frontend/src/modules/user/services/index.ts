import { ComputedRef, Ref } from 'vue';
import {
  API_USER,
  API_USER_PASSWORD,
  TGetUsersDTO,
  TGetUsersQueryDTO,
  TGetUserDTO,
  TPostUserDTO,
  TPostUserDataDTO,
  TUpdateUserDTO,
  TUpdateUserDataDTO,
  TDeleteUserDTO,
  TUpdateUserPasswordDTO,
  TUpdateUserPasswordDataDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const userService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_USER, page],
      queryFn: async () => {
        const params: TGetUsersQueryDTO = { page: page.value };

        const { data } = await api.get<TGetUsersDTO>(API_USER, { params });

        return data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined> | Ref<string | undefined>) => {
    return useQuery({
      queryKey: [API_USER, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetUserDTO>(`${API_USER}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TPostUserDataDTO) => {
        const { data } = await api.post<TPostUserDTO>(API_USER, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (formData: TUpdateUserDataDTO) => {
        const { data } = await api.patch<TUpdateUserDTO>(`${API_USER}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  updatePassword: (options: object, id?: string) => {
    return useMutation({
      mutationKey: [API_USER_PASSWORD, id],
      mutationFn: async (password: TUpdateUserPasswordDataDTO) => {
        if (!id) return null;

        const { data } = await api.patch<TUpdateUserPasswordDTO>(`${API_USER_PASSWORD}/${id}`, password);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_USER],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteUserDTO>(`${API_USER}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
