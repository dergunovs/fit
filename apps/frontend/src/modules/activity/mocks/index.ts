import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
