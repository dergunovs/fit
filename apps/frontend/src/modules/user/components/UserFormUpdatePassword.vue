<template>
  <UiSpoiler :title="t('updatePassword')" v-model="isShowUpdatePassword">
    <UiFlex gap="12">
      <UiInput
        v-model="newPassword"
        :placeholder="t('password6Symbols')"
        isPassword
        data-test="user-form-update-password-input"
      />

      <UiButton
        :isDisabled="newPassword.length < 6"
        isNarrow
        @click="mutateUpdatePassword({ password: newPassword, id: props.id })"
        data-test="user-form-update-password-button"
      >
        {{ t('update') }}
      </UiButton>
    </UiFlex>
  </UiSpoiler>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { UiButton, UiFlex, UiInput, UiSpoiler, toast } from 'mhz-ui';

import { userService } from '@/user/services';
import { useTI18n } from '@/common/composables';

interface IProps {
  id: string;
}

interface IEmit {
  updated: [];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t } = useTI18n();

const isShowUpdatePassword = ref(false);
const newPassword = ref('');

const { mutate: mutateUpdatePassword } = userService.updatePassword({
  onSuccess: async () => {
    toast.success(t('passwordUpdated'));
    newPassword.value = '';
    isShowUpdatePassword.value = false;

    emit('updated');
  },
});
</script>
