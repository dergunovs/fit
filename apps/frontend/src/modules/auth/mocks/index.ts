import { ref } from 'vue';
import { vi } from 'vitest';
import {
  TLocale,
  TPostAuthConfirmTokenDataDTO,
  TPostAuthConfirmTokenDTO,
  TPostAuthLoginDataDTO,
  TPostAuthLoginDTO,
  TPostAuthRegisterDataDTO,
  TPostAuthRegisterDTO,
  TPostAuthResetPasswordDataDTO,
  TPostAuthResetPasswordDTO,
  TPostAuthSetupDataDTO,
  TPostAuthSetupDTO,
} from 'fitness-tracker-contracts';

import { authService } from '@/auth/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import * as authComposables from '@/auth/composables';
import { IOnSuccess } from '@/common/interface';
import { USER_FIXTURE } from '@/user/fixtures';

const mockIsAdmin = ref(true);
const mockIsAuthChecked = ref(true);

const spyUseAuthCheck = vi
  .spyOn(authComposables, 'useAuthCheck')
  .mockReturnValue({ user: ref(USER_FIXTURE), isAdmin: ref(mockIsAdmin), isAuthChecked: ref(mockIsAuthChecked) });

const spyLogin = vi.fn();
const spySetup = vi.fn();
const spyRegister = vi.fn();
const spyConfirmToken = vi.fn();
const spyResetPassword = vi.fn();

const mockOnSuccess: IOnSuccess = {
  check: undefined,
  login: undefined,
  setup: undefined,
  register: undefined,
  confirmToken: undefined,
  resetPassword: undefined,
};

const spyCheckAuth = vi.spyOn(authService, 'check').mockImplementation(() => mockQueryReply(USER_FIXTURE));

vi.spyOn(authService, 'login').mockImplementation(
  (options: { onSuccess?: (user: TPostAuthLoginDTO) => Promise<void> }) => {
    if (options.onSuccess) mockOnSuccess.login = options.onSuccess;

    return mockMutationReply<TPostAuthLoginDTO, TPostAuthLoginDataDTO>(spyLogin);
  }
);

vi.spyOn(authService, 'setup').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.setup = options.onSuccess;

  return mockMutationReply<TPostAuthSetupDTO, TPostAuthSetupDataDTO>(spySetup);
});

vi.spyOn(authService, 'register').mockImplementation((_lang: TLocale, options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.register = options.onSuccess;

  return mockMutationReply<TPostAuthRegisterDTO, TPostAuthRegisterDataDTO>(spyRegister);
});

vi.spyOn(authService, 'confirmToken').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.confirmToken = options.onSuccess;

  return mockMutationReply<TPostAuthConfirmTokenDTO, TPostAuthConfirmTokenDataDTO>(spyConfirmToken);
});

vi.spyOn(authService, 'resetPassword').mockImplementation(
  (_lang: TLocale, options: { onSuccess?: () => Promise<void> }) => {
    if (options.onSuccess) mockOnSuccess.resetPassword = options.onSuccess;

    return mockMutationReply<TPostAuthResetPasswordDTO, TPostAuthResetPasswordDataDTO>(spyResetPassword);
  }
);

export {
  spyUseAuthCheck,
  spyCheckAuth,
  spyLogin,
  spySetup,
  spyRegister,
  spyConfirmToken,
  spyResetPassword,
  mockOnSuccess,
  mockIsAdmin,
};
