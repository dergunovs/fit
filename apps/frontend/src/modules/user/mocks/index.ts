import { vi } from 'vitest';
import {
  TDeleteUserDTO,
  TGetUsersDTO,
  TPostUserDataDTO,
  TPostUserDTO,
  TUpdateUserDataDTO,
  TUpdateUserDTO,
} from 'fitness-tracker-contracts';

import { userService } from '@/user/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { USER_FIXTURE, USERS_FIXTURE } from '@/user/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetUser = vi.spyOn(userService, 'getOne').mockReturnValue(mockQueryReply(USER_FIXTURE));

const mockGetUsersData: TGetUsersDTO = { data: USERS_FIXTURE, total: USERS_FIXTURE.length };

const spyGetUsers = vi.spyOn(userService, 'getMany').mockImplementation(() => mockQueryReply(mockGetUsersData));

const spyCreateUser = vi.fn();
const spyUpdateUser = vi.fn();
const spyDeleteUser = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(userService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostUserDTO, TPostUserDataDTO>(spyCreateUser);
});

vi.spyOn(userService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateUserDTO, TUpdateUserDataDTO>(spyUpdateUser);
});

vi.spyOn(userService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteUserDTO, string>(spyDeleteUser);
});

export { spyGetUser, spyGetUsers, spyCreateUser, spyUpdateUser, spyDeleteUser, mockOnSuccess, mockGetUsersData };
