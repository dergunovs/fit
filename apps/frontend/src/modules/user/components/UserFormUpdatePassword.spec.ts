import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import UserFormUpdatePassword from './UserFormUpdatePassword.vue';

import { wrapperFactory } from '@/common/test';
import { mockOnSuccess, spyUpdateUserPassword } from '@/user/mocks';
import { spyToastSuccess } from '@/common/mocks';

const input = dataTest('user-form-update-password-input');
const button = dataTest('user-form-update-password-button');

let wrapper: VueWrapper<InstanceType<typeof UserFormUpdatePassword>>;

const NEW_PASSWORD = '123456';
const ID = 'gfhrtj46j4y54hhrt';

beforeEach(() => {
  wrapper = wrapperFactory(UserFormUpdatePassword, { id: ID });
});

enableAutoUnmount(afterEach);

describe('UserFormUpdatePassword', () => {
  it('exists', () => {
    expect(wrapper.findComponent(UserFormUpdatePassword)).toBeTruthy();
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('disables button when password is less than 6 characters', async () => {
    expect(wrapper.find(button).attributes('isdisabled')).toBe('true');

    await wrapper.findComponent(input).setValue('12345');

    await wrapper.find(button).trigger('click');

    expect(wrapper.find(button).attributes('isdisabled')).toBe('true');

    await wrapper.findComponent(input).setValue(NEW_PASSWORD);

    expect(wrapper.find(button).attributes('isdisabled')).toBe('false');
  });

  it('updates password', async () => {
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyUpdateUserPassword).toBeCalledTimes(0);

    await wrapper.findComponent(input).setValue('123456');

    await wrapper.find(button).trigger('click');

    await mockOnSuccess.updatePassword?.();

    expect(spyToastSuccess).toBeCalledTimes(1);
    expect(spyUpdateUserPassword).toBeCalledTimes(1);
    expect(spyUpdateUserPassword).toBeCalledWith({ password: NEW_PASSWORD, id: ID });
  });
});
