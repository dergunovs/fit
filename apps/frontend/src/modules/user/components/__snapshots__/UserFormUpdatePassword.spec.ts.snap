// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserFormUpdatePassword > matches snapshot 1`] = `
"<div title="Обновить пароль" modelvalue="false">
  <div gap="12"><input type="text" modelvalue="" placeholder="Пароль от 6 символов" ispassword="" data-test="user-form-update-password-input"><button isdisabled="true" isnarrow="" data-test="user-form-update-password-button">Обновить</button></div>
</div>"
`;
